airbyte-cdk/python/airbyte_cdk/connector_builder/migrator/source.py

```diff
@@ -0,0 +1,166 @@
+#
+# Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+#
+
+import json
+import logging
+import os
+import re
+from enum import Enum
+from typing import Any, Dict, List, Optional
+
+import yaml
+import yaml.emitter
+import yaml.representer
+import yaml.resolver
+import yaml.serializer
+from airbyte_cdk.sources.declarative.models.declarative_component_schema import AuthFlow, Spec
+
+logger = logging.getLogger("migrator.source")
+
+
+class SourceRepository:
+
+    def __init__(self, base_path: str) -> None:
+        self._base_path = base_path
+
+    def merge_spec_inside_manifest(self, source_name: str) -> None:
+        spec = self._fetch_spec_outside_manifest(source_name)
+        if spec:
+            manifest_path = self._manifest_path(source_name)
+            with open(os.path.join(manifest_path)) as f:
+                manifest = yaml.safe_load(f.read())
+            if "spec" in manifest:
+                # It should be fine to ignore since the CATs test `test_match_expected` ensure that both are the same
+                logger.info(
+                    f"Source {source_name} has both spec file and spec within manifest. Will keep manifest as-is and delete the spec file"
+                )
+            else:
+                manifest["spec"] = spec.dict(exclude_none=True)
+
+            with open(manifest_path, 'w') as outfile:
+                yaml.dump(manifest, outfile, sort_keys=False, Dumper=_ManifestDumper)
+            self._delete_spec_outside_manifest(source_name)
+        else:
+            logger.debug(f"Source {source_name} does not have a spec outside the manifest.yaml")
+
+    def _fetch_spec_outside_manifest(self, source_name: str) -> Optional[Spec]:
+        json_spec_path = os.path.join(self._python_package_path(source_name), "spec.json")
+        yaml_spec_path = os.path.join(self._python_package_path(source_name), "spec.yaml")
+
+        if os.path.exists(json_spec_path):
+            with open(json_spec_path) as json_file:
+                return self._assemble_spec(json.load(json_file))
+        elif os.path.exists(yaml_spec_path):
+            with open(yaml_spec_path) as yaml_file:
+                return self._assemble_spec(yaml.safe_load(yaml_file))
+        else:
+            return None
+
+    @staticmethod
+    def _assemble_spec(spec: Dict[str, Any]) -> Spec:
+        return Spec(
+            type="Spec",
+            connection_specification=spec["connectionSpecification"],
+            documentation_url=spec.get("documentationUrl", None),
+            advanced_auth=AuthFlow(**spec["advanced_auth"]) if "advanced_auth" in spec else None
+        )
+
+    def _delete_spec_outside_manifest(self, source_name: str) -> None:
+        json_spec_path = os.path.join(self._python_package_path(source_name), "spec.json")
+        yaml_spec_path = os.path.join(self._python_package_path(source_name), "spec.yaml")
```

Comment:
just a thought: We do have some code duplication where we're checking multiple times whether json_spec or the yaml_spec exist. What we could do to simplify this a bit is:
```
def _get_spec_path(self, source_name: str) -> (Optional[str], Optional[str]):
  json_spec_path = os.path.join(self._python_package_path(source_name), "spec.json")
  yaml_spec_path = os.path.join(self._python_package_path(source_name), "spec.yaml")

  // maybe also validate there are not both?

  return (
    json_spec_path if os.path.exists(json_spec_path) else None,
    yaml_spec_path if os.path.exists(yaml_spec_path) else None
  )
```

We would do this first and pass those values into flows like the remove _fetch_spec_outside_manifest, we know that the presence of the path set implies it exists instead of reading from the OS again. Not a big deal or blocking, but that came to mind that we could figure out which spec we're dealing earlier

Reply:
Good idea, will update soon!